include/device.h

```diff
@@ -568,6 +568,33 @@ int device_required_foreach(const struct device *dev,
  */
 __syscall const struct device *device_get_binding(const char *name);
 
+/**
+ * @brief Retrieve the device structure for a driver by name.
+ *
+ * This is the implementation underlying device_from_name(), without
+ * the overhead of a syscall wrapper.
+ *
+ * @param name device name to search for.  A null pointer, or a pointer to an
+ * empty string, will cause NULL to be returned.
+ *
+ * @return pointer to device structure; NULL if the name is not
+ * acceptable or no device is associated with the name.
+ */
+const struct device *z_device_from_name(const char *name);
+
+/**
+ * @brief Retrieve the device structure for a driver by name.
+ *
+ * This differs from device_get_binding() in that it does not verify
+ * that the device is ready to use.
+ *
+ * @param name device name to search for.  A null pointer, or a pointer to an
+ * empty string, will cause NULL to be returned.
+ *
+ * @return pointer to device structure; NULL if not found or cannot be used.
+ */
+__syscall const struct device *device_from_name(const char *name);
```

Comment:
`device_lookup_by_name` is better suited to be the syscall and `device_from_name` the static function.

Reply:
But we already have `device_from_handle`, so it's more consistent to do it this way.